### ISSUE TITLE
Usage with OfflineAudioContext

### ISSUE THREAD

**Issue by this-fifo** [danigb/smplr#88]
Hi @danigb ! — Hope you've been doing well. I have been using your library in my projects and really like the performance and so far all the custom Soundfonts I've used are working well.

That said, recently I have ran into an issue that I'm struggling a lot to understand and I wanted to ask you a question to see if you have tried or know about this.

What I am trying to do is to download a song through using OfflineAudioContext for faster processing. The issue I'm having is that when I plug in `smplr` it doesn't work despite everything being connected right. I either get no sound at all, or, strangely, I hear the first second but then it goes silent.

I am initializing `smplr` by passing in the context as a variable — which can be either an Offline or a Real time one — I then schedule the events using the `Tone.js` transport.

When the context is a real time one it plays all the notes, but in the offline mode it doesn't.

I feel like I have hit a wall here so I decided to ask if you have tried using `smplr` with an OfflineAudioContext before and how did you do it if so.

Any help or advice is much appreciated, thank you so much! 🙇‍♂️🙏

## More details below

To try and add more context, I use a callback to play the notes kinda like this where `player` is either a `Soundfont` or `Soundfont2Sampler` instance
```js
(time, value) => {
      value.notes.forEach((note) => {
            player.start({
                  time,
                  note,
                  velocity,
                  duration,
                  loop: true,
            })
      })
}
```

Then I schedule events on the offline transport and the resulting AudioBuffer this way
```js
return Tone.Offline(
  async (context) => {
    context.transport.seconds = 0
    context.transport.bpm.value = state.bpm
    playFunctionsAndNotes.forEach(([playFunction, values]) => {
      values.forEach((sequence) => {
        context.transport.schedule((time) => {
          playFunction(time, sequence)
        }, sequence.time)
      })
    })
    context.transport.start(0)
  },
  SONG_DURATION, // 2mins
  CHANNELS, // 2
  SAMPLE_RATE, // 44100 (same as the soundfont buffers)
)
```

Here's another interesting remark, in my array of play functions and notes, some of the play functions have `Tone.Synth` instruments instead of `smplr` based ones. These get rendered just fine into the resulting `AudioBuffer` through the `OfflineAudioContext` but the `Soundfont | Soundfont2Sampler` based ones strangely only play for the first 1 second and their first notes. 😭

In short, I hear 2 mins of correctly played `Tone.Synth` sounds but only 1 second of `Soundfont | Soundfont2Sampler` sounds.

I also made sure that, by the time it reaches that code, the instruments were loaded through `await new Soundfont( ... ).load`

**Comment by danigb** [comment]
Hey!

I don't have time to test it, but the reason could be related with the internal scheduler. I've added an option to disable it: https://github.com/danigb/smplr/blob/main/CHANGELOG.md

Can you try with `disableScheduler: true` ? (you need to upgrade to latest version first)

**Comment by this-fifo** [comment]
Hey, thank you so much for getting back to me and for the change. That did the trick!

However, I am now having a weird behaviour, the rendered audio from `smplr` is off tempo. My song is on 160 bpm but the `smplr` instrument track is on 120 bpm for some reason.

I am still debugging why the tempo is off, but for now I am super happy that disabling the scheduler made the audio work!

Would you mind sharing a bit more about how the internals/scheduler work or point me to it? You're the best! 🙌

**Comment by danigb** [comment]
> Hey, thank you so much for getting back to me and for the change. That did the trick!

Glad to hear it!

> However, I am now having a weird behaviour, the rendered audio from smplr is off tempo. My song is on 160 bpm but the smplr instrument track is on 120 bpm for some reason.

Since you're using an external scheduler (tonejs transport) I guess is related to it. Not sure what could be the cause without diving into the code. If I'm correct, with the scheduler disabled, smplr will start notes at the specific time given. Maybe the time from the scheduler is different from the web audio api? Just tinking out loud...

> Would you mind sharing a bit more about how the internals/scheduler work or point me to it? 

Well, basically some people were using smplr to schedule a whole score (like >2k notes) at a time, and that made Web Audio API crash or go very slow. So the internal scheduler what is does is: when a note is started with more than 0.2s in advance, it adds to a waiting queue and only sends it to web audio api when is about to play. That also allows to use smplr to drive the UI (because the `onStart` callback will be called just some ms before the start time).

Of course, that doesn't play well with offline usage (I never thought about this use case before).

You can find more details here: https://github.com/danigb/smplr/blob/main/src/player/queued-player.ts
